provisioning/miniprov/hare_mp/cdf.py

```diff
@@ -33,6 +33,8 @@ def _create_node_descriptions(self) -> List[NodeDesc]:
         nodes: List[NodeDesc] = []
         conf = self.provider
         server_dict: Dict[str, str] = conf.get('cluster>server_nodes')
+        if not server_dict:
+            raise MissingKey(Text('cluster>server_nodes'))
```

Comment:
Why use `Text`?

Exceptions help to distinguish the nature of the problem (exception type) and help to provide human-friendly text that explains what was wrong (message).

In this case, I guess `raise MissingKey('Required key cluster>server_nodes not found')` will make way more sense.

Second point is.. Can't we move this `if` and this raising into the logic of `conf.get()`? Semantics of the call would be "I return the value or I raise the exception if the value is not present". That would simplify the code around, no extra checks will need to be added then.

Reply:
1. In this case, I guess raise MissingKey('Required key cluster>server_nodes not found') will make way more sense.

> Final output will be 'Error performing configuration. Missing ConfStore key ("cluster>server_nodes")_'.  Config() function which handles MissingKey exception will output 'Missing ConfStore key (%s)', key_error' by replacing 'key_error' with specific key.

If we move raising logic to conf.get() then it will always raise exception when key is not preset.
We may not always want to raise exception when key is missing, sometimes we can just take default value if key is missing.

We can still achieve this with moving raising logic to conf.get() and handling exception at caller but cases where we don't want to raise exception if key is missing needs to be handled separately.

@knekrasov  Let me know your inputs

Reply:
@SwapnilGaonkar7 `Text` type is a kind of [DTO type](https://en.wikipedia.org/wiki/Data_transfer_object) (or [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree), to be more precise) to express something in Dhall language. `'Missing ConfStore key (%s)', key_error` will invoke `__str__()` from `Text` class which will return the wrapped string. If you wrap a string with `Text` and then only thing that you do is implicitly unwrap it again into a string, then I don't see why `Text` is needed. From logical point of view, it makes no sense to me.

> We can still achieve this with moving raising logic to conf.get() and handling exception at caller but cases where we don't want to raise exception if key is missing needs to be handled separately.

On the contrary, I would like it to raise the exception whenever the key is incorrect. It will make the caller code more straightforward, there is no fear that somebody just forgets to check if the value is None (and thus will not notice a mistype in the key name). Please see [fail-fast](https://en.wikipedia.org/wiki/Fail-fast) principle for more details.

If by some reason the programmer will want another behavior, (s)he can add one more method like `if_key_exists` (or just implement `in` operator) but nowadays we don't have such a need.

Reply:
Small update:
In my current PR I already realized the need of some values to be None (network_interface_type can be None). So I implemented it like this:

```python
class ValueProvider:
    def get(self, key: str, allow_null: bool = False) -> Any:
        ret = self._raw_get(key)
        if ret is None and not allow_null:
            raise RuntimeError(f'ConfStore key {key} not found')
        return ret

    def _raw_get(self, key: str) -> str:
        raise NotImplementedError()


class ConfStoreProvider(ValueProvider):
    def __init__(self, url: str):
        self.url = url
        conf = ConfStore()
        conf.load('hare', url)
        self.conf = conf

    def _raw_get(self, key: str) -> str:
        return self.conf.get('hare', key)
```

So when your PR is merged, I'll just update raw `RuntimeError` with your `MissingKey` type in my code above.

Reply:
Just for reference: #1548

Reply:
@knekrasov Addressed your review comments. Please have a look